system/jlib/jqueue.tpp

```diff
@@ -547,4 +547,137 @@ public:
                                     while (x--) 
 
 
+template <class BASE>
+class RingBufferOf
+{
+    typedef RingBufferOf<BASE> SELF;
+    BASE **ptrs;
+    unsigned headp;
+    unsigned tailp;
+    unsigned max;
+    unsigned num;
+
+    virtual void releaseBuffer(BASE *e){};
+    virtual void releaseBufferAll()
+    {
+        while (num)
+        {
+            releaseBuffer(ptrs[headp]);
+            ++headp;
+            --num;
+            if (headp==max) headp=0;
+        }
+    };
+
+public:
+
+    inline RingBufferOf() : ptrs(NULL), headp(0), tailp(0), max(0), num(0){};
+
+    virtual ~RingBufferOf() { clear(); }
+
+    inline void clear()
+    {
+        releaseBufferAll();
+        delete [] ptrs;
+        ptrs=NULL;
+        headp=0;
+        tailp=0;
+        max=0;
+        num=0;
+    }
+    inline void ensure(unsigned newSize)
+    {
+        if (newSize==max) return;
+
+/*        if (max)
+        {
+            if (newSize < max)
+            {
+                if (headp<=tailp)
+                {
+                    memmove(ptrs,ptrs+headp, headp*sizeof(BASE *));
+                    tailp -= headp;
+                    headp = 0;
+                } else
+                {
+                }
+            }
+        }
+        ptrs = (BASE **)realloc(ptrs,newSize*sizeof(BASE *));
+        assertex(!newSize || ptrs);
+
+        if (max)
+        {
+            if (newSize > max)
+            {
+                if (tailp<headp)
+                {
+                    // Create space for the nodes from headp to end
+                    memmove(ptrs+max-headp,ptrs, max-headp);
+                    memmove(ptrs, headp+(max-headp), max-headp);
+                    tailp += (max-headp);
+                    headp = 0;
+                    max = newSize;
+                }
+            }
+        }*/
+        //clear();
+        max = newSize;
+        ptrs = new BASE* [max];
```

Comment:
We will probably want to allocate this out of roxie memory, so an alloc/realloc could be a member of BASE template.